File: 167_two_sum_II_input_array_is_sorted.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        for (int i = 0; i < nums.length; i++) {
            int lo = i + 1, hi = nums.length - 1, diff = target - nums[i];
            while (lo <= hi) {
                int mid = lo + (hi - lo) / 2;
                if (nums[mid] == diff) {
                    return new int[] {i + 1, mid + 1};
                } else if (nums[mid] > diff) {
                    hi = mid - 1;
                } else {
                    lo = mid + 1;
                }
            }
        }
        return null;
    }
}

File: python/167_two_sum_II_input_array_is_sorted.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # apply binary search
        left = 0
        right = len(nums) - 1
        while left < right:
            mid_idx = left + (right - left) // 2
            if nums[left] + nums[right] < target:
                if nums[mid_idx] + nums[right] < target:
                    left = mid_idx + 1
                else:
                    left += 1
            elif nums[left] + nums[right] > target:
                if nums[mid_idx] + nums[left] > target:
                    right = mid_idx
                else:
                    right -= 1
            else:
                return [left + 1, right + 1]
        return []
